### PR TITLE
rowlock and snapshot code refactoring

### PR DESCRIPTION
**Pull request by ldennis**
from #361

### CODE REVIEW

File: db/config.c [hunk 1]
```diff
@@ -1040,10 +1040,15 @@ static int read_lrl_option(struct dbenv *dbenv, char *line, void *p, int len)
     } else if (tokcmp(tok, ltok, "enable_snapshot_isolation") == 0) {
         bdb_attr_set(dbenv->bdb_attr, BDB_ATTR_SNAPISOL, 1);
         gbl_snapisol = 1;
+        gbl_new_snapisol = 1;
+        gbl_new_snapisol_asof = 1;
+        gbl_new_snapisol_logging = 1;
+        logmsg(LOGMSG_INFO, "Enabled snapshot isolation (default newsi)\n");
     } else if (tokcmp(tok, ltok, "enable_new_snapshot") == 0) {
```
Comment (markhannum) [inline db/config.c]: Yeah - agreed

File: bdb/phys.c [hunk 2]
```diff
@@ -255,18 +255,63 @@ int phys_dta_add(bdb_state_type *bdb_state, tran_type *logical_tran,
...
+        if (rc) {
+            logmsg(LOGMSG_ERROR, "%s failed to allocate rlptr rc %d\n",
+                   __func__, rc);
+            goto done;
+        }
+
+        if (log_compare(&last_regop_lsn, &logical_tran->last_regop_lsn))
+            did_commit = 1;
+
+        rc = bdb_lock_row_write_getlock(bdb_state, logical_tran, -1, genid,
```
Comment (markhannum) [inline bdb/phys.c]: I was convinced that we didn't need to get a writelock for dta file inserts .. my thinking was that given that genids are unique, it's impossible to have collisions here.

File: bdb/ll.c [hunk 3]
```diff
@@ -342,11 +344,29 @@ int ll_dta_del_rowlocks(bdb_state_type *bdb_state, tran_type *tran, int rrn,
...
+                            0, &odh, &freeptr);
+            if (rc) {
+                if (freeptr)
+                    free(freeptr);
+                dbcp->c_close(dbcp);
+                goto done;
+            }
+
+            /* Verify updateid for rowlocks */
+            if (tran->logical_tran && dtafile == 0 &&
```
Comment (ldennis) [inline bdb/ll.c]: use read_odh() or bdb_cposition()

File: db/record.c [hunk 4]
```diff
@@ -1989,7 +1989,11 @@ int del_record(struct ireq *iq, void *trans, void *primkey, int rrn,
     rc = dat_del(iq, trans, rrn, genid);
     if (iq->debug)
         reqprintf(iq, "DEL RRN %d GENID 0x%llx RC %d", rrn, genid, rc);
-    if (rc != 0) {
+    if (rc == ERR_VERIFY && bdb_is_rowlocks_transaction(trans)) {
+        *opfailcode = OP_FAILED_VERIFY;
+        retrc = rc;
+        goto err;
+    } else if (rc != 0) {
         *opfailcode = OP_FAILED_INTERNAL + ERR_DEL_DTA;
```
Comment (ldennis) [inline db/record.c]: *opfailcode = rc == ERR_VERIFY ? OP_FAILED_VERIFY : OP_FAILED_INTERNAL + ERR_DEL_DTA;

File: bdb/phys.c [hunk 5]
```diff
@@ -496,7 +501,7 @@ int phys_dta_upd(bdb_state_type *bdb_state, int rrn,
         goto done;
     }
 
-    if (dtafile == 0) {
+    if (dtafile == 0 && !logical_tran->single_physical_transaction) {
```
Comment (ldennis) [inline bdb/phys.c]: if (dtafile == 0 && gbl_rowlocks) {

File: bdb/rowlocks.c [hunk 6]
```diff
@@ -1852,6 +1886,9 @@ int abort_logical_transaction(bdb_state_type *bdb_state, tran_type *tran,
         LOGCOPY_32(&rectype, logdta.data);
     }
 
+    if (tran->physical_tran)
+        bdb_tran_abort_phys(bdb_state, tran->physical_tran);
```
Comment (ldennis) [inline bdb/rowlocks.c]: remove

**Comment by riverszhang89** [comment]
I'd prefer `logmsg(LOGMSG_DEBUG, ...)` to `#ifdef DEBUG printf(...) #endif`.

**Pull request merged**